website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md

```diff
@@ -0,0 +1,592 @@
+
+```graphql
+extend type Query {
+  users(ids: [ID!]! @is(field: "id")): [User!]
+}
+```
+
+The `@is` directive allows us to specify that the argument on our field `users` is semantically the same as the output field `id` on ther returning `User` type. Since `ids` is a list and it returns a list of users we can now infer that this field allows us to batch fetch users by user ids.
+
+## Requirements
+
+Where subgraph directives really become necessary is with requirements. Requirements are things we need from somewhere else befor we can execute a field.
+
+Let's say we have the following schema:
+
+```graphql
+type Product {
+  sku: String!
+  name: String!
+  dimension: ProductDimension
+}
+```
+
+Also let's say we have a subgraph that is able to calculate a delivery estimate for a product.
+
+```graphql
+type Product {
+  deliveryEstimate(zip: String!, width: Float!, height: Float!): Int!
+}
+```
+
+In order to calculate the shiping estimate in our shipping subgraph we need the ZIP code, the width and height of the product. This dimenson of the product is actually available in the product catalog service that holds all the information about the product itself.
+
+In this case we want to create a public api for our consumer where we only have to pass in the ZIP code to the `deliveryEstimate` field on the `Product` type on our Fusion graph.
+
+```graphql
+type Product {
+  sku: String!
+  name: String!
+  dimension: ProductDimension
+  deliveryEstimate(zip: String!): Int!
+}
+```
+
+We can express this by using the `@require` directive and refer to the required information relatively to the `Product` type.
+
+```graphql
+type Product {
+  deliveryEstimate(
+    zip: String!,
+    width: Float! @require(field: "dimension { width }"),
+    height: Float! @require(field: "dimension { height }")) : Int!
+}
+```
+
+We also could design that slightly differently and introduce a input to our subgraph that represents the required data we need.
+
+```graphql
+input ProductDimensionInput {
+  width: Float!,
+  height: Float!
+}
+
+type Product {
+  deliveryEstimate(
+    zip: String!,
+    dimension: ProductDimensionInput! @require(field: "dimension")): Int!
+}
+```
+
+The outcome will stay the same and we get this nice API for our user and the query planner will resolve the requiered data under the hood.
+
+![Banana Cake Pop - Query Plan Viewer](bcp-2.png)
+
+Again, this brings actual clarity to your subgraph as the field is very clear what it needs and also becomes easily testable.
+
+## Reshaping things
+
+When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not reall need to be on a product type as the argument has clear requirements which are expressed by it's field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself it could very well be exposed through the `Query` type. At least in the context of our subgraph.
```

Comment:
```suggestion
When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not really need to be on a product type as the argument has clear requirements which are expressed by its field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself, it could be exposed through the `Query` type. At least in the context of our subgraph.
```